java/impl/gcp/src/main/java/co/worklytics/psoxy/GcpApiDataRequestHandler.java

```diff
@@ -0,0 +1,258 @@
+        this.requestHandler = requestHandler;
+        this.gcpEnvironment = gcpEnvironment;
+        this.apiModeConfig = apiModeConfig;
+        this.envVarsConfigService = envVarsConfigService;
+        this.googleIdTokenVerifierFactory = googleIdTokenVerifierFactory;
+        this.objectMapper = objectMapper;
+    }
+
+    @SneakyThrows
+    public void service(HttpRequest request, HttpResponse response) {
+
+        // check if request is invocation via PubSub
+        // as of 2025-08-19, appears as 'APIs-Google; (+https://developers.google.com/webmasters/APIs-Google.html)'
+        Boolean userAgentIsPubSub = request.getFirstHeader(HttpHeaders.USER_AGENT)
+            .map(userAgent -> 
+                userAgent.contains(gcpEnvironment.getGoogleApisUserAgent()) || // in practice, this is how invocation for Cloud Run v2 via PubSub is coming
+                userAgent.contains(gcpEnvironment.getPubSubUserAgent()))
+            .orElse(false);
+
+        if (userAgentIsPubSub) {
+            if (envVarsConfigService.isDevelopment()) {
+                log.log(Level.INFO, "PubSub push invocation detected");
+            }
+
+            // potentially async case - PubSub push invocation
+            if (request.getFirstHeader(GcpEnvironment.PUBSUB_DELIVERY_ATTEMPT_HEADER).map(Integer::parseInt).orElse(-1) > MAX_ASYNC_ATTEMPTS) {
+                log.log(Level.SEVERE, "Max PubSub delivery attempts exceeded, dropping message");
+                response.setStatusCode(org.apache.hc.core5.http.HttpStatus.SC_NO_CONTENT, "Max delivery attempts exceeded");
+                return;
+            }
+
+            // verify auth to ENSURE is legit PubSub push callback
+            verifyAuthentication(request);
+
+            PubSubPushBody pubSubPushBody = objectMapper.readerFor(PubSubPushBody.class)
+                .readValue(new ByteArrayInputStream(request.getInputStream().readAllBytes()));
+
+            if (envVarsConfigService.isDevelopment()) {
+                log.log(Level.INFO, "PubSub push body: " + pubSubPushBody);
+            }
+
+            HttpEventRequest originalRequest = objectMapper.readerFor(HttpEventRequestDto.class)
+                .readValue(Base64.getDecoder().decode(pubSubPushBody.getMessage().getData().getBytes(StandardCharsets.UTF_8)));
+
+            ApiDataRequestHandler.ProcessingContext processingContext = objectMapper.readerFor(ApiDataRequestHandler.ProcessingContext.class)
+                .readValue(pubSubPushBody.getMessage().getAttributes().get(ApiDataRequestViaPubSub.MessageAttributes.PROCESSING_CONTEXT.getStringEncoding()));
+
+            processingContext = processingContext.toBuilder()
+                .async(true)
+                .build();
+
+
+            HttpEventResponse genericResponse = requestHandler.handle(originalRequest, processingContext);
+
+            // TODO: probably DO NOT want body/etc here, right??
+            fillGcpResponseFromGenericResponse(response, genericResponse);
```

Comment:
I'd move this to a `handleAsyncCase` method.
Minor: probably would do the sync case first, as should be more frequent.

Reply:
Ok for encapsulating

as far as order, my feeling is to avoid negation operator unless you need.